### DISCUSSION TITLE
Sticky layer functionality

### DISCUSSION THREAD

**Discussion by TimoFreiberg** [manna-harbour/miryoku#15] [Q&A] [answered]
Sometimes I'm eating something while watching Youtube or reading an article, and many keys are quite hard to press with just one hand.
I think it would be nice to have a way to keep layers enabled, I'm thinking of pressing e.g. SPC+Z to keep the NAVR layer enabled, maybe until I press SPC or SPC+Z again?
No idea how easily something like that is doable in QMK, I'm just starting out :)

**Comment by manna-harbour** [comment]
Thanks for creating this issue!

Miryoku is designed to take advantage of the increased ergonomics possible when using both hands, but at the expense of casual use, as it requires both hands to be used.

The work-around for situations that require one hand use would be to add a custom layer and activate it with a layer toggle.  I plan on including some examples but I imagine they'd really need to be tailored to each user, so it's a low priority.

But your suggestion of just locking the current layer would probably be enough in many cases, should be quite easy to add, and wouldn't affect the rest of the layout.  I'll look into it and see if I can add something like that.  Would you be interested in testing it?

**Comment by TimoFreiberg** [comment]
Thanks for the reply, yeah I would be interested in testing :)

**Comment by manna-harbour** [comment]
Great, I'll let you know when I have something!

**Comment by manna-harbour** [comment]
It'd be nice to be able to lock and unlock with a single key, but there is no layer lock in QMK (although I might add one if this turns out to be useful).  TO and TG can't be used to lock the layer within a held LT.  DF will work to change which layer you return to after a LT hold release, and another DF can be used to get back to the base layer.  This works but it requires one key to give the effect of locking and another to unlock.

I've implemented it in [this branch](https://github.com/manna-harbour/qmk_firmware/tree/miryoku-layerlock-test) (as a manual change to the generated c files, not in the org source). In a sublayer, on the same hand as the held LT, tap the bottom row pinkie key to lock, and tap the thumb key normally used to activate the layer to unlock.  Please let me know if it works and if it's useful, and if you have any suggestions.

**Reply by manna-harbour** [reply]
> TO and TG can't be used to lock the layer within a held LT. DF will work to change which layer you return to after a LT hold release, and another DF can be used to get back to the base layer. This works but it requires one key to give the effect of locking and another to unlock.

Likewise, functions equivalent to `TO` or `TG` can't be used for the same reason: `ACTION_TAP_DANCE_LAYER_MOVE`, `layer_move`, `layer_on`, `layer_state_set`. Use `default_layer_set` instead. Also note that `DF` can't be used in `TD`.

**Comment by TimoFreiberg** [comment]
Thanks a lot!
I think you made a copy-paste mistake in the bottom row of the NSSL layer though :sweat_smile: 
The layer locking seems to work fine, I'm going to test drive it for a few days :)

**Comment by manna-harbour** [comment]
Thanks, fixed now!

**Comment by manna-harbour** [comment]
See #52.

**Comment by hanslenze** [comment]
This feature works great! It's nice to be able to scroll through big documents and webpages without holding the thumb button to keep  the navigation/mouse layer activated. The layer hold toggle is in a good position to reach it comfortably and consistently with a single hand.
I've also added the following buttons to the top left on the mouse layer for comfortable web surfing: RCS(KC_TAB), KC_WBAK, KC_WREF, KC_WFWD, LCTL(KC_TAB), With this addition I'm able to stay in the mouse layer much longer while browsing. I've chosen not to add "new tab" and "close tab" because most of the time those actions coincide with typing on the base layer.

**Reply by manna-harbour** [reply]
Thanks for your report!  Did you also add a layer lock for single hand activation, or only on the same layer requiring both hands to lock?

**Reply by hanslenze** [reply]
I've used your proposed method of same-side-pinky on most layers. That works great because it doesn't interfere with the opposite hand on the same layer. The layer switching hand usually has a few modifiers but nothing on the top and bottom row. 
I'll add my mouse layer below for clarification:

| col1  | col2   | col3  | col4 | col5 | col6 | col7 | col8 | col9 | col10 |
|----------------|-------------|-------------|-------------|--------------------|-------------|------------|-------------|-----------|-------|
| RCS(KC_TAB), |  KC_WBAK, |  KC_WREF, | KC_WFWD, |  LCTL(KC_TAB),       | X_NU,      | KC_BTN1, | KC_MS_U,  | KC_BTN2, | KC_WH_U, |
| KC_LSFT,        | KC_LGUI,   |  KC_LALT,  | KC_LCTL,     | X_NA,                    | X_NU,      | KC_MS_L, | KC_MS_D,  | KC_MS_R,  | KC_WH_D, |
| DF(MOUR),    | KC_ALGR,  |  X_NA,      |  X_NA,        |  X_NA,                    | X_NU,      | KC_WH_L, | KC_WH_D, | KC_WH_U, | KC_WH_R, |
| X_NP,             | X_NP,        |  X_NA,      |  X_NA,         | DF(BASE),              | KC_BTN1, | KC_BTN3, | KC_BTN2,   | X_NP,        | X_NP |

edit: codeblock garbles my neatly formatted text.

**Comment by Shahabaz-Bagwan** [comment]
There is toggle function for layers in ZMK and QMK it will be good for the above mentioned use case.

**Reply by manna-harbour** [reply]
Unfortunately it can't be used to lock the current layer, see https://github.com/manna-harbour/miryoku/discussions/15#discussioncomment-1558247.  But it might help with one-hand lock, as that will be on a different layer.

**Comment by manna-harbour** [comment, marked as answer]
Layer lock for current and opposite layers has been added. See https://github.com/manna-harbour/miryoku/tree/master/docs/reference#additional-features.
